image-rs/src/verity.rs

```diff
@@ -0,0 +1,102 @@
+    let device_block_size = verity_option.blocksize.parse::<u64>().unwrap();
+    let hash_block_size = verity_option.hashsize.parse::<u64>().unwrap();
+    let hash_start_block: u64 =
+        (device_num_blocks * device_block_size + hash_block_size - 1) / hash_block_size;
+
+    // verity parameters: <version> <data_device> <hash_device> <data_blk_size> <hash_blk_size>
+    // <blocks> <hash_start> <algorithm> <root_hash> <salt>
+    // version: on-disk hash version
+    //     0 is the original format used in the Chromium OS.
+    //     1 is the current format that should be used for new devices.
+    // data_device: device containing the data the integrity of which needs to be checked.
+    //     It may be specified as a path, like /dev/vdX, or a device number, major:minor.
+    // hash_device: device that that supplies the hash tree data.
+    //     It is specified similarly to the data device path and is the same device in the function of create_verity_device.
+    //     The hash_start should be outside of the dm-verity configured device size.
+    // data_blk_size: The block size in bytes on a data device.
+    // hash_blk_size: The size of a hash block in bytes.
+    // blocks: The number of data blocks on the data device. 
+    // hash_start: offset, in hash_blk_size blocks, from the start of hash_device to the root block of the hash tree.
+    // algorithm: The cryptographic hash algorithm used for this device. This should be the name of the algorithm, like "sha256".
+    // root_hash: The hexadecimal encoding of the cryptographic hash of the root hash block and the salt.
+    // salt: The hexadecimal encoding of the salt value.
+    let verity_params = format!(
+        "1 {} {} {} {} {} {} {} {} {}",
+        source_device_path,
+        source_device_path,
+        verity_option.blocksize,
+        verity_option.hashsize,
+        verity_option.blocknum,
+        hash_start_block,
+        verity_option.hashtype,
+        verity_option.hash,
+        "-",
+    );
+    // Mapping table in device mapper: <start_sector> <size> <target_name> <target_params>: 
+    // <start_sector> is 0 
+    // <size> is size of device in sectors
+    // <target_name> is name of mapping target, here "verity" for dm-verity
+    // <target_params> is parameters for verity target
+    let verity_table = vec![(0, device_num_blocks, "verity".into(), verity_params)];
+
+    dm.device_create(&verity_name, None, opts)?;
+    dm.table_load(&id, verity_table.as_slice(), opts)?;
+    dm.device_suspend(&id, opts)?;
+    Ok(format!("{}{}", "/dev/mapper/", &verity_option.hash))
+}
+
+pub fn get_verity_device_name(verity_options: &str) -> Result<String> {
+    let parsed_data = decode_verity_options(verity_options)?;
+    Ok(parsed_data.hash)
+}
+
+pub fn decode_verity_options(verity_options: &str) -> Result<DmVerityOption> {
+    let decoded = base64::decode(verity_options)?;
+    let parsed_data = serde_json::from_slice::<DmVerityOption>(&decoded)?;
```

Comment:
Please help to validate parsed DmVerityOption here.

Reply:
Thanks, I'll update it.